src/main/java/it/eng/idsa/dataapp/web/rest/DataControllerHttpHeader.java

```diff
@@ -44,39 +47,80 @@ public DataControllerHttpHeader(MessageUtil messageUtil) {
 			logger.info("Payload is empty");
 		}
 
-		String requestMessageType = httpHeaders.getFirst("IDS-Messagetype");
+		HttpHeaders responseHeaders = createResponseMessageHeaders(httpHeaders);
+		String responsePayload = null;
+		
+		if (!("ids:RejectionMessage".equals(responseHeaders.get("IDS-Messagetype").get(0)))) {
+			responsePayload = messageUtil.createResponsePayload(httpHeaders);
+		} else {
+			responsePayload = "Rejected message";
+		}
+		
+		if (responsePayload.contains("IDS-RejectionReason")) {
+			httpHeaders.put("IDS-Messagetype", Util.asList("ids:RejectionMessage"));
+			httpHeaders.put("IDS-RejectionReason", Util.asList(responsePayload.substring(responsePayload.indexOf(":")+1)));
+			responseHeaders = createResponseMessageHeaders(httpHeaders);
+			responsePayload = "Rejected message";
+		}
 
 		return ResponseEntity.ok().header("foo", "bar")
-				.headers(createResponseMessageHeaders(requestMessageType))
+				.headers(responseHeaders)
 				.header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
-				.body(messageUtil.createResponsePayload(requestMessageType));
+				.body(responsePayload);
 	}
 
-	private HttpHeaders createResponseMessageHeaders(String requestMessageType) {
+	private HttpHeaders createResponseMessageHeaders(HttpHeaders httpHeaders) {
+		String requestMessageType = httpHeaders.getFirst("IDS-Messagetype");
 		HttpHeaders headers = new HttpHeaders();
 
 		DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
 		Date date = new Date();
 		String formattedDate = dateFormat.format(date);
 
 		String responseMessageType = null;
+		String rejectionReason = null;
 
-		if ("ids:ContractRequestMessage".equals(requestMessageType)) {
+		switch (requestMessageType) {
+		case "ids:ContractRequestMessage":
 			responseMessageType = ContractAgreementMessage.class.getSimpleName();
-		} else {
-			responseMessageType = ArtifactResponseMessage.class.getSimpleName();
+			break;
+
+		case "ids:ArtifactRequestMessage":
+			if (httpHeaders.containsKey("IDS-TransferContract")
+					&& !("https://contract.com/12".equals(httpHeaders.get("IDS-TransferContract").get(0)))
```

Comment:
Are following URL's the same like DEFAULT_CONTRACT_AGREEMENT and other?

Reply:
Sharp eyes, I missed this one, thanks!